Grafana-CLI: Fix broken database commands for deb/rpm packages

gotjosh opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our [`CONTRIBUTING.md`](https://github.com/grafana/grafana/blob/master/CONTRIBUTING.md) guide.
2. Ensure you have added or ran the appropriate tests for your PR.
3. If it's a new feature or config option it will need a docs update. Docs are under the docs folder in repo root.
4. If the PR is unfinished, mark it as a draft PR.
5. Rebase your PR if it gets out of sync with master
6. Name your PR as `<FeatureArea>: Describe your change`. If it's a fix or feature relevant for changelog describe the user impact in the title. The PR title is used in the changelog for issues marked with `add to changelog` label. 
-->

**What this PR does / why we need it**:

When our users install the *nix packed version of grafana (via RPM or deb), tendency is to use the services and scripts installed as part of the package for grafana-server. These leverage the default configuration options by specifying several default paths.

This introduces a similar approach for the grafana-cli binary. We expose it through a wrapper to ensure a proper configuration is in place. 

To enable that, we tuck away into the `/usr/share/grafana` (default grafana home) directory the original binary (grafana-cli) and then use a bash script named grafana-cli as the wrapper.

In addition to that, this PR also:

- Makes the `--config` and `--homepath` flags global. Which allow us to significantly simplify the wrapper logic. 
- Adds a new  `--optionsString` which allow us to override default configuration options as a string. A thing to note is that this was and is still supported by passing these arguments as trailing args to the command, with the particularity that the trailing arguments take precedence over the flag to allow for backward compatibility. That means:

```bash
$ ./bin/darwin-amd64/grafana-cli -d --optionsString="cfg:default.paths.data=/dev/nullX cfg:default.paths.logs=/dev/nullX" --config=/dev/null admin reset-admin-password 1234 cfg:default.paths.plugins=/dev/nullY cfg:default.paths.logs=/dev/nullY
INFO[06-20|14:33:00] Config loaded from                       logger=settings file=/Users/josueabreu/go/src/github.com/grafana/grafana/conf/defaults.ini
INFO[06-20|14:33:00] Config loaded from                       logger=settings file=/dev/null
INFO[06-20|14:33:00] Config overridden from command line      logger=settings arg="default.paths.data=/dev/nullX"
INFO[06-20|14:33:00] Config overridden from command line      logger=settings arg="default.paths.logs=/dev/nullY"
INFO[06-20|14:33:00] Config overridden from command line      logger=settings arg="default.paths.plugins=/dev/nullY"
INFO[06-20|14:33:00] Path Home                                logger=settings path=/Users/josueabreu/go/src/github.com/grafana/grafana
INFO[06-20|14:33:00] Path Data                                logger=settings path=/dev/nullX
INFO[06-20|14:33:00] Path Logs                                logger=settings path=/dev/nullY
INFO[06-20|14:33:00] Path Plugins                             logger=settings path=/dev/nullY
INFO[06-20|14:33:00] Path Provisioning                        logger=settings path=/Users/josueabreu/go/src/github.com/grafana/grafana/conf/provisioning
INFO[06-20|14:33:00] App mode production                      logger=settings
INFO[06-20|14:33:00] Connecting to DB                         logger=sqlstore dbtype=sqlite3
INFO[06-20|14:33:00] Starting DB migration                    logger=migrator

Error: New password is too short

NAME:
   Grafana cli admin reset-admin-password - reset-admin-password <new password>

USAGE:
   Grafana cli admin reset-admin-password [arguments...
```
- Logs the paths configured as part of the command run to provide some feedback for the user if the `--debug` global flag is used

<!--
*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
-->

gotjosh committed:
Feature: Introduce a grafana-cli wrapper

When our users install the *nix packed version of grafana, tendency is to use the services and scripts installed as part of the package for grafana-server. These leverage the default configuration options by specifying the several default paths.

This introduces a similar approach for the grafana-cli binary. We exposed it through a wrapper to ensure a proper configuration is in place. To enable that, we add the .real suffix to the original binary (grafana-cli.real) and then use a bash script named grafana-cli as the wrapper.

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/grafana/grafana?pullRequest=17596) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/grafana/grafana?pullRequest=17596) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/grafana/grafana?pullRequest=17596) it.</sub>

gotjosh commented on build.go:
```diff
@@ -391,7 +400,7 @@ func createPackage(options linuxPackageOptions) {
 	}
 
-	// add dependenciesj
+	// add dependencies
```
Assumed this was a typo and not a way to spell dependencies in Swedish 😆

bergquist commented:
The "data-migration" command need the homepath and config as well. All commands that interact with the database needs those to figure out where the grafana.ini config is.

bergquist reviewed (approved):
LGTM

bergquist reviewed (changes requested):
It doesnt works as expected for me in the hosted-grafana image. Im currently trying to figure out why.

gotjosh commented:
@bergquist I'm happy to investigate if you can provide me with a bit more detail. I tested within our circleCI image and it works as expected.

bergquist commented on packaging/wrappers/grafana-cli.wrapper:
```diff
@@ -0,0 +1,58 @@
+COMMAND=$1
+SUBCOMMAND=$2
+
+# We special case admin commands as is they are the only ones the that supports --homepath and --config as flags.
```
could we pass those args regardless if the command needs them? Just to make the script simpler.

gotjosh replied:
Sadly 😢 Since the configuration overrides are not flags the CLI ends up thinking they are commands.

bergquist commented on pkg/cmd/grafana-cli/commands/commands.go:
```diff
@@ -19,12 +18,15 @@ func runDbCommand(command func(commandLine utils.CommandLine, sqlStore *sqlstore
 			Config:   cmd.String("config"),
 			HomePath: cmd.String("homepath"),
-			Args:     flag.Args(),
+			Args:     context.Args().Tail(),
```
I haven't really had time to figure out exactly why. But `context.Args().Tail()` does not work for me when executing `grafana-cli admin data-migration encrypt-datasource-passwords`. I'm *guessing* its related to args count but I dont know right now.

 Given we just 

```go
flag.Parse()
flag.Args()
```
in `grafana-server`. perhaps we should stick to that here as well?

gotjosh replied:
It's exactly that - is related to the position of the arguments. It's the main downside to the wrapper approach, we'll always need to keep it up to date with whichever commands we add.

My last commit goes in a slightly different direction - with this branch you should have no problem using the `grafana-cli` directly as is the original binary + the fix in the first commit. The wrapper now has an explicit `.wrapper` suffix.

That being said, I'm happy to add a bit more logic to the wrapper to take into account sub-commands. However, it feels a bit strange that `data-migration` sub-command only has 1 command underneath it, could it be that the plan is to add more of these in the future? If not, I'd be in favour of deleting the third level to bring it inline to the `admin` command - the downside? Backwards compatibility unless we create an alias.

gotjosh replied:
My last commit removes the level of indentation - if we decide to move in a different direction I'm happy to take it out.

bergquist commented:
The related issue is in our private repo. But you should be able to see a reference to is on this issue.

gotjosh commented:
I ended up going in a slightly different direction in my last commit and would love some feedback on what do people reckon. Instead of trying to be implicit for our users and "mask" the fact that we're doing something behind the scenes. I opted to introduce the wrapper in a more explicit manner.

The approach moves away from `grafana-cli` as a wrapper and `grafana-cli.real` as the original binary, to `grafana-cli.wrapper` as the wrapper and `grafana-cli` as the original binary. In my investigation, I saw at least one example [[1]](http://manpages.ubuntu.com/manpages/cosmic/man8/aide.wrapper.8.html) which follows this approach that I think sets precedent.

I think this works around the current challenges described in the PR description where our users end up confused about "`grafana-cli.real`" and it is 100% backwards compatible as there's no change in the original binary. However, I do understand this would need additional documentation (perhaps as part of the package installation) to make our users aware that it exists.

xlson commented:
I'm skeptical about giving the bash-script a different name than `grafana-cli`, seems like it would make it harder to use for very little gain. Correctly configuring `grafana-cli` without the wrapper is sufficiently complicated that I have a hard time seeing that being something many users will want to do. As mentioned, we could solve the naming problem by putting the real grafana-cli somewhere else but with the same name.

An alternative to the bash wrapper if the wrapper + cli-parsing gets too complicated is to double back and implement the configuration reading in the wrapper directly.

Let's discuss this when we all have time.

aocenas commented on packaging/wrappers/grafana-cli:
```diff
@@ -0,0 +1,43 @@
+#! /usr/bin/env bash
+
+# Wrapper for the grafana-cli binary
```
Cool I wanted to do this for some time 👍

aocenas commented on packaging/wrappers/grafana-cli:
```diff
@@ -0,0 +1,43 @@
+fi
+
+OPTS="--homepath=${GRAFANA_HOME} \
+      --config=${CONF_FILE} \
```
So with this I assume we do not need to do the `if dev { change pluginsDir }` we have now in the CLI right? It should pick up the one in config? In the config though the setting is `plugins = data/plugins` for example so is this picked somehow? Also this will need to be appended to homepath but that is also different when doing development so not sure if that is handled somehow.

gotjosh replied:
> So with this I assume we do not need to do the if dev { change pluginsDir } we have now in the CLI right?

Sorry @aocenas 🙈, not sure I follow the questions. I'm not entirely sure where `if dev { change pluginsDir }` is coming from.

>  It should pick up the one in config? In the config though the setting is plugins = data/plugins for example so is this picked somehow? 

That is correct, because of:

```bash
# overwrite settings from default file
if [ -f "$DEFAULT" ]; then
  . "$DEFAULT"
fi
``` 

We load the settings from the default file and override with whatever we have here. 


> Also this will need to be appended to homepath but that is also different when doing development so not sure if that is handled somehow.

Not sure I follow that last question but if what we mean by development is us developing in our local machines this wouldn't affect it as the wrapper only affects RMP and deb packaging.

aocenas replied:
Sorry this was probably a bit cryptic. In the CLI there is this part https://github.com/grafana/grafana/blob/be66ed9dabcc7d8d76510b39330d523e3efe1da2/pkg/cmd/grafana-cli/utils/grafana_path.go#L11 where we try to get the plugin dir when you are on dev environment and actually figure out if we are in dev environment. It does not work very well and the diff that @bergquist ping you on in my PR is my trying to fix that for second time.

So I think there could be some more robust solution like getting it from the config.ini where you already need need to setup it correctly so that grafana itself knows where to find installed plugins. At the same time I wanted it to work no matter what is the current cwd (like if I run ./bin/grafana-cli or ./grafana-cli or ./bin/darwin-amd64/grafana-cli ideally I would like the same results).  

> he wrapper only affects RMP and deb packaging.

In this case though this probably isn't going to help so we will need to fix the paths on dev in other PR anyway.

gotjosh replied:
Agreed! Sorry, but this wouldn't really help - I'm pretty sure we could think of something but it feels like it sits outside of the scope of this PR.

gotjosh committed:
Make the config and homepath flags global

gotjosh committed:
Introduce optionsStrings as a global flag

This flag allows us to pass configuration overrides as a string.

The string follows the convention of configuration arguments separated by a space e.g. "cfg:default.paths.data=/dev/nullX cfg:default.paths.logs=/dev/nullX"

Also, it is backwards compatible with similar the previous configuration method through tailing arguments. Tailing arguments take presedence over the configuration options string.

gotjosh committed:
Only log configuration information in debug mode
